ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,606 @@
+ * and thus it is able to capture the pre-app2app state.
+ *
+ */
+
+static dr_emit_flags_t
+drstatecmp_app2app_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                         bool translating, OUT void **user_data)
+{
+    /* Allocate space for user_data. */
+    drstatecmp_user_data_t *data = (drstatecmp_user_data_t *)dr_thread_alloc(
+        drcontext, sizeof(drstatecmp_user_data_t));
+    memset(data, 0, sizeof(*data));
+    *user_data = (void *)data;
+
+    /* Determine whether the basic block is free of side effects. */
+    data->side_effect_free_bb = true;
+    for (instr_t *inst = instrlist_first_app(bb); inst != NULL;
+         inst = instr_get_next_app(inst)) {
+        if (drstatecmp_may_have_side_effects_instr(inst)) {
+            data->side_effect_free_bb = false;
+            return DR_EMIT_DEFAULT;
+        }
+    }
+
+    /* Current bb is side-effect free. Save a copy of this pre-app2app bb. */
+    data->pre_app2app_bb = instrlist_clone(drcontext, bb);
+    return DR_EMIT_DEFAULT;
+}
+
+/****************************************************************************
+ * ANALYSIS PHASE
+ *
+ * The analysis phase determines which copy of each side-effect bb should be used as the
+ * golden copy for comparison. There are two options: i) pre-app2app-phase copy, where
+ * the code just contains the original app instructions, and ii) post-app2app-phase copy.
+ * The first option can catch bugs in app2app passes and it is selected unless any of
+ * original app instructions requires emulation (true emulation). In that case, the
+ * pre-app2app code is not executable, whereas the post-app2app code contains emulation
+ * code and thus can be used for state checks with re-execution. Cases that do
+ * not require the emulation sequence for re-execution include instruction
+ * refactoring that simplify instrumentation but do not correspond to true
+ * emulation (e.g., drutil_expand_rep_string() and drx_expand_scatter_gather()).
+ *
+ */
+
+static dr_emit_flags_t
+drstatecmp_analyze_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                         bool translating, void *user_data)
+{
+    drstatecmp_user_data_t *data = (drstatecmp_user_data_t *)user_data;
+
+    if (!data->side_effect_free_bb)
+        return DR_EMIT_DEFAULT;
+
+    for (instr_t *inst = instrlist_first(bb); inst != NULL; inst = instr_get_next(inst)) {
+        if (drmgr_is_emulation_start(inst)) {
+            emulated_instr_t emulated_inst;
+            drmgr_get_emulated_instr_data(inst, &emulated_inst);
+            if (!(emulated_inst.flags & DR_EMULATE_INSTR_ONLY)) {
+                /* Emulation sequence required for re-execution and golden bb copy should
+                 * be the post-app2app copy.
+                 */
+                instrlist_clear_and_destroy(drcontext, data->pre_app2app_bb);
+                data->golden_bb_copy = instrlist_clone(drcontext, bb);
```

Comment:
Best to have a test case for this path

Reply:
Should be exercised now with the fuzz app.